notebooks/linear_regression.livemd

```diff
@@ -0,0 +1,610 @@
+# Linear Regression in Practice
+
+```elixir
+Mix.install(
+  [
+    {:scholar, github: "elixir-nx/scholar"},
+    {:explorer, github: "elixir-nx/explorer"},
+    {:nx, github: "elixir-nx/nx", sparse: "nx", override: true},
+    {:exla, github: "elixir-nx/nx", sparse: "exla", override: true},
+    {:rustler, ">= 0.0.0", optional: true},
+    {:req, "~> 0.3.3"},
+    {:vega_lite, "~> 0.1.6"},
+    {:kino_vega_lite, "~> 0.1.7"},
+    {:kino, "~> 0.8.0"}
+  ],
+  config: [
+    nx: [default_defn_options: [compiler: EXLA]]
+  ]
+)
+```
+
+## Introduction
+
+In the livebook, we will cover the typical use cases of linear regression on practical examples.
+
+```elixir
+alias VegaLite, as: Vl
+require Explorer.DataFrame
+```
+
+<!-- livebook:{"branch_parent_index":0} -->
+
+## Linear Regression on Synthetic Data
+
+```elixir
+use Nx.Defn.Kernel
+```
+
+Before we dive into real-life use cases of linear regression, we start with a simpler one. We will generate data with a linear pattern and then use `Scholar.Linear.LinearRegression` to compute regression.
+
+Firstly, we generate the data which simulates the function $f(x) = 3x + 4$ with added uniform, zero-mean noise. `Nx.Random.uniform` creates a tensor with a given shape and type.
+
+```elixir
+key = Nx.Random.key(42)
+size = 100
+{uniform_x, new_key} = Nx.Random.uniform(key, shape: {size, 1}, type: :f32)
+{uniform_y, _} = Nx.Random.uniform(new_key, shape: {size, 1}, type: :f32)
+x = 2 * uniform_x
+y = 3 * x + 4 + (uniform_y - Nx.mean(uniform_y))
+```
+
+Now let's plot the generated points.
+
+```elixir
+df = Explorer.DataFrame.new(x: Nx.to_flat_list(x), y: Nx.to_flat_list(y))
+
+Vl.new(
+  title: [
+    text: "Scatterplot of Generated Data",
+    offset: 20
+  ],
+  width: 500,
+  height: 500
+)
+|> Vl.data_from_values(df)
+|> Vl.mark(:circle)
+|> Vl.encode_field(:x, "x",
+  type: :quantitative,
+  scale: [domain: [-0.05, 2.05]],
+  axis: [grid: false]
+)
+|> Vl.encode_field(:y, "y", type: :quantitative, scale: [domain: [2.5, 12]], axis: [grid: false])
+```
+
+For a regression task, we will use the `Scholar.Linear.LinearRegression` module.
+
+```elixir
+model = Scholar.Linear.LinearRegression.fit(x, y)
+```
+
+As we can see, the coefficient is almost 3.0, and the intercept is nearly 4.0. Those are decent estimations. They are not exactly equal to 3.0 and 4.0 because we introduce noise to our samples.
```

Comment:
```suggestion
As we can see, the coefficient is almost 3.0, and the intercept is nearly 4.0. Those are decent estimations. They are not exactly equal to 3.0 and 4.0 because we introduced noise to our sample generation.
```